src/aleph/http/multipart.clj

```diff
@@ -162,42 +170,87 @@
+  ReferenceCounted
+  (refCnt [_]
+    (.refCnt raw-http-data))
+  (retain [_]
+    (.retain raw-http-data))
+  (retain [_ increment]
+    (.retain raw-http-data increment))
+  (^ReferenceCounted touch [_]
+    (.touch raw-http-data))
+  (^ReferenceCounted touch [_ ^Object hint]
+    (.touch raw-http-data hint))
+  (release [_]
+    (.release raw-http-data))
+  (release [_ decrement]
+    (.release raw-http-data decrement)))
+
 (defmulti http-data->map
   (fn [^InterfaceHttpData data]
     (.getHttpDataType data)))
 
 (defmethod http-data->map InterfaceHttpData$HttpDataType/Attribute
   [^Attribute attr]
   (let [content (.getValue attr)]
-    {:part-name (.getName attr)
-     :content content
-     :name nil
-     :charset (-> attr .getCharset .toString)
-     :mime-type nil
-     :transfer-encoding nil
-     :memory? (.isInMemory attr)
-     :file? false
-     :file nil
-     :size (count content)}))
+    (map->MultipartChunk
+     {:part-name (.getName attr)
+      :content content
+      :name nil
+      :charset (-> attr .getCharset .toString)
+      :mime-type nil
+      :transfer-encoding nil
+      :memory? (.isInMemory attr)
+      :file? false
+      :file nil
+      :size (count content)
+      :raw-http-data attr})))
 
 (defmethod http-data->map InterfaceHttpData$HttpDataType/FileUpload
   [^FileUpload data]
   (let [memory? (.isInMemory data)]
-    {:part-name (.getName data)
-     :content (when memory?
-                (bs/to-input-stream (netty/acquire (.content data))))
-     :name (.getFilename data)
-     :charset (-> data .getCharset .toString)
-     :mime-type (.getContentType data)
-     :transfer-encoding (.getContentTransferEncoding data)
-     :memory? memory?
-     :file? true
-     :file (when-not memory? (.getFile data))
-     :size (.length data)}))
+    (map->MultipartChunk
+     {:part-name (.getName data)
+      :content (when memory?
+                 (bs/to-input-stream (netty/acquire (.content data))))
+      :name (.getFilename data)
+      :charset (-> data .getCharset .toString)
+      :mime-type (.getContentType data)
+      :transfer-encoding (.getContentTransferEncoding data)
+      :memory? memory?
+      :file? true
+      :file (when-not memory? (.getFile data))
+      :size (.length data)
+      :raw-http-data data})))
 
 (defn- read-attributes [^HttpPostRequestDecoder decoder parts]
-  (while (.hasNext decoder)
-    (s/put! parts (http-data->map (.next decoder)))))
+  (d/loop []
+    (if-not (.hasNext decoder)
+      (d/success-deferred true) ;; go for another chunk of body
+      (let [^InterfaceHttpData data (.next decoder)]
+        (if (nil? data)
+          ;; this probably could happen only in case of
+          ;; simultaneous access to the decoder object...
+          (d/success-deferred true)
+          (d/chain'
+           (s/put! parts (http-data->map data))
+           (fn [succeed?]
+             (if succeed?
+               (do
+                 (.removeHttpDataFromClean decoder data)
```

Comment:
Why are we telling it not to clean up the file again?

Reply:
Here is the rationale behind the idea. [1]

> Reimplemented public API using io.netty.util.ReferenceCounted and relying on netty/release for each consumed chunk. I played a bit with the version of API that exposes a single .close method for the entire decoder object and found that it lacks flexibility as I cannot deallocate memory from a specific chunk (potentially freeing a lot of space as soon as possible) while all other chunks are still in processing... So, I think that releasing them one-by-one is the best option in terms of resources usage even tho' it might be a bit confusing from the first glance (well, nothing special if you have some experience with Netty/C/C++/Rust etc, but probably not something you're doing in Clojure on a daily basis). /cc @gsnewmark @ztellman

Basically, it transfers the responsibility of cleaning up the resources from `HttpPostRequestDecoder#destroy` to the user so each part/chunk can be removed one per one.   

> But if so, shouldn't we only call this when .isInMemory is true?

The same applies when it's a file. 
When `.release` (and thus `.deallocate`) [2] is called on a `AbstractMixedHttpData`, it cleans the associated file [3].

[1] : https://github.com/clj-commons/aleph/pull/432#issuecomment-457755670
[2] : https://github.com/netty/netty/blob/4.1/codec-http/src/main/java/io/netty/handler/codec/http/multipart/AbstractMixedHttpData.java#L101
[3] : https://github.com/netty/netty/blob/4.1/codec-http/src/main/java/io/netty/handler/codec/http/multipart/AbstractDiskHttpData.java#L281